File: 2022/05/javascript/main.mjs
import fs from "node:fs";
import readline from "node:readline";

// --- Constants ---
const INPUT_FILE = "../input/input.txt";
const CRATES_REGEX = /(\[\w\]|\s{3})\s?/g;
const STEPS_REGEX = /\d+/g;

let queues1 = [];
let queues2 = [];
let setup = true;

const stream = fs.createReadStream(INPUT_FILE);
const reader = readline.createInterface({ input: stream, crlDelay: Infinity });
reader.on("line", line => {
	// --- Parse Initial crates setup ---
	if (setup) {
		if (line.indexOf("[") >= 0) {
			let index = 0;
			for (let token of line.match(CRATES_REGEX)) {
				if (queues1.length - 1 < index) {
					queues1.push([]);
					queues2.push([]);
				}
				let c = token.charAt(1);
				if (c !== " ") {
					queues1[index].push(c);
					queues2[index].push(c);
				}
				index++;
			}
		}

		// --- An empty line marks ends of setup ---
		if (line.length === 0) setup = false;
	} else {
		// --- Parse script ---
		let [count, from, to] = line.match(STEPS_REGEX);
		from--;
		to--;

		// --- Puzzle 1 ---
		for (let i = 0; i < count; i++) {
			let crate = queues1[from].splice(0, 1)[0];
			queues1[to].unshift(crate);
		}

		// --- Puzzle 2 ---
		for (let i = count - 1; i >= 0; i--) {
			let crate = queues2[from].splice(i, 1)[0];
			queues2[to].unshift(crate);
		}
	}

});
await new Promise(res => reader.once("close", res));

console.log("Part 1. Crates on the top of each stack: ", queues1.map(q => q[0]).join(""));
console.log("Part 2. Crates on the top of each stack: ", queues2.map(q => q[0]).join(""));

File: 2022/05/java/Main.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Main {
	private static final String INPUT_FILE = "../input/input.txt";
	private static final Pattern cratesPattern = Pattern.compile("(\\[\\w\\]|\\s{3})\\s?");
	private static final Pattern stepsPattern = Pattern.compile("\\d+");

	public static void main(String[] args) throws IOException {
		List<List<Character>> queues1 = new ArrayList<>();;
		List<List<Character>> queues2 = new ArrayList<>();;

		boolean setup = true;
		try (BufferedReader reader = new BufferedReader(new FileReader(INPUT_FILE))) {
			String line;
			while ((line = reader.readLine()) != null) {
				// --- Parse Initial crates setup ---
				if (setup) {
					if (line.indexOf('[') >= 0) {
						Matcher matcher = cratesPattern.matcher(line);
						int index = 0;
						while(matcher.find()) {
							Character token = line.charAt(matcher.start() + 1);
							if (queues1.size() - 1 < index) {
								queues1.add(new ArrayList<>());
								queues2.add(new ArrayList<>());
							}
							if (token != ' ') {
								queues1.get(index).add(token);
								queues2.get(index).add(token);
							}
							index++;
						}
					}

					// --- An empty line marks ends of setup ---
					if (line.length() == 0) {
						setup = false;
					}
					continue;
				}

				// --- Parse script ---
				Matcher matcher = stepsPattern.matcher(line);
				matcher.find();
				int count = Integer.parseInt(matcher.group());
				matcher.find();
				int from = Integer.parseInt(matcher.group()) - 1;
				matcher.find();
				int to = Integer.parseInt(matcher.group()) - 1;

				// --- Puzzle 1 ---
				for (int i = 0; i < count; i++) {
					Character crate = queues1.get(from).remove(0);
					queues1.get(to).add(0, crate);
				}

				// --- Puzzle 2 ---
				for (int i = count - 1; i >= 0; i--) {
					Character crate = queues2.get(from).remove(i);
					queues2.get(to).add(0, crate);
				}
			}
		}

		System.out.print("Part 1. Crates on the top of each stack: ");
		for (List<Character> queue: queues1) System.out.print(queue.get(0));
		System.out.println();

		System.out.print("Part 2. Crates on the top of each stack: ");
		for (List<Character> queue: queues2) System.out.print(queue.get(0));
		System.out.println();
	}
}
